dashboard/config/scripts_json/coursea-2019.script_json

```diff
@@ -0,0 +1,2385 @@
+      },
+      "named_level": null,
+      "bonus": null
+    },
+    {
+      "level_names": [
+        "courseA_playLab_events4_2019"
+      ],
+      "lesson_key": "Events in Play Lab",
+      "lesson_group_key": "csf_events",
+      "script_name": "coursea-2019",
+      "seeding_id": "f1d6e3fecc69da14017513493708fb53",
+      "chapter": 98,
+      "position": 5,
+      "assessment": null,
+      "properties": {
+        "progression": "Mini-project: Jorge the Dog"
+      },
+      "named_level": null,
+      "bonus": null
+    },
+    {
+      "level_names": [
+        "courseA_playLab_events5_2019"
+      ],
+      "lesson_key": "Events in Play Lab",
+      "lesson_group_key": "csf_events",
+      "script_name": "coursea-2019",
+      "seeding_id": "c614d8d248bd03868ce514b5a761f81e",
+      "chapter": 99,
+      "position": 6,
+      "assessment": null,
+      "properties": {
+        "progression": "Mini-project: Jorge the Dog"
+      },
+      "named_level": null,
+      "bonus": null
+    },
+    {
+      "level_names": [
+        "courseA_playLab_events6_2019"
+      ],
+      "lesson_key": "Events in Play Lab",
+      "lesson_group_key": "csf_events",
+      "script_name": "coursea-2019",
+      "seeding_id": "d46dc472609411f91841ca249e18a50a",
+      "chapter": 100,
+      "position": 7,
+      "assessment": null,
+      "properties": {
+        "progression": "Mini-project: Jorge the Dog"
+      },
+      "named_level": null,
+      "bonus": null
+    },
+    {
+      "level_names": [
+        "courseA_playLab_events7_2019"
+      ],
+      "lesson_key": "Events in Play Lab",
+      "lesson_group_key": "csf_events",
+      "script_name": "coursea-2019",
+      "seeding_id": "e1621e8eece898df88b88424a8779cbf",
+      "chapter": 101,
+      "position": 8,
+      "assessment": null,
+      "properties": {
+        "progression": "Mini-project: Jorge the Dog"
+      },
+      "named_level": null,
+      "bonus": null
+    },
+    {
+      "level_names": [
+        "courseA_playlab_events_challenge2_2019"
+      ],
+      "lesson_key": "Events in Play Lab",
+      "lesson_group_key": "csf_events",
+      "script_name": "coursea-2019",
+      "seeding_id": "1f9ff0524a6e65c9fa5582f5e9a5b7c2",
+      "chapter": 102,
+      "position": 9,
+      "assessment": null,
+      "properties": {
+      },
+      "named_level": null,
+      "bonus": true
+    },
+    {
+      "level_names": [
+        "courseA_playlab_events_challenge1_2019"
+      ],
+      "lesson_key": "Events in Play Lab",
+      "lesson_group_key": "csf_events",
+      "script_name": "coursea-2019",
+      "seeding_id": "1d804a966ac671158cbe6900f46c71bb",
+      "chapter": 103,
+      "position": 10,
+      "assessment": null,
+      "properties": {
+      },
+      "named_level": null,
+      "bonus": true
+    }
+  ],
+  "levels_script_levels": [
+    {
+      "level_key": "GoingPlacesSafely-unplugged",
+      "script_level_seeding_id": "d00eabec25b4d0f36eb0d52763ce8955"
```

Comment:
Hm, probably nothing, this was just the way I initially thought to do it. We could probably change it.

Reply:
to continue this discussion... it looks like each entry in `script_levels` each contains a list of `level_names` already, so it seems like you could achieve the same result here by doing a second pass over script_levels, and creating a levels_script_levels object for each level name. This would let you delete the LevelScriptLevel model, its serializer, and the script_levels seeding_id. I would lean toward making this simplification but I'm fine with you deciding based on all that you know about this space.

I do like that what you have here proves out a strategy for serializing something which does not have a unique key across environments. However, I'll need to use something similar to this strategy for Activity and ActivitySection, since those will not have special `key` fields, just the environment-specific `id`, so those could also live on as the example of this strategy.

Reply:
@davidsbailey thanks for bringing up the new models - I was thinking that for every new model we add, we should have a rule that it has a unique index that isn't the `id` column. This just simplifies things a lot, and makes it easier to get good performance, compared to for `script_levels` where the uniquely identifying info requires data from another table.

If there's no natural candidate for a unique index to use, I'd suggest adding a `key` column which is just a GUID, and generate one for every object on creation. It could be exposed through the API or not, whichever is more convenient for the API, but at a minimum it would be used for seeding, and make that simpler.

We actually could go this route and backfill GUIDs for ScriptLevel, and if it proves to be too big a blocker in simplifying the code it'll start to be tempting.

Reply:
Thanks for going into this Winter . Initially i was thinking that new models would omit the cross-environment `key` column and have only the environment-specific `id` column, and follow the example you had here where script_level had GUIDs generated just for seeding. the goal was to not store unnecessary data. that seems like it would work, but one downside is that then the GUIDs would not be stable, and they could (would?) change every time we re-serialize an object. is that why you moved away from this strategy?

This seems like a good enough reason, so I'm fine with adding a `key` column on the new models which contains a GUID. this seems much simpler than, for example, having a human-readable key that is only unique within the scope of each lesson, since then we have to do more work to combine the human-readable key with the lesson's unique key in order to be able to use it.

Assuming we go with this strategy, back-filling script levels with globally unique `keys` sounds great to me. it seems better to rally around a single, simple strategy rather than have to combine a bunch of existing data in a complicated way just to avoid having an explicit `key`. although the `key` will be opaque (since it is a GUID), it will probably be easier to understand while debugging rather than trying to understand/parse a more complex key.

Reply:
I would still be open to removing LevelsScriptLevels from the serialized format, though I think having a model for it adds a nice abstraction for the seeding logic, especially since I've added the `seeding_key` abstraction. Would still be open to thinking about removing it in a followup PR, but I don't think it's necessary for this change.

As for backfilling script levels with unique keys - it would be a good simplification but I'd prefer to look into that in a followup PR / leave it as optional.